File: platform/cli-go/internal/commands/undo.go
package commands

import (
	"fmt"
	"math"
	"os/exec"
	"time"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

func newUndoCmd() *cobra.Command {
	var list bool

	cmd := &cobra.Command{
		Use:   "undo",
		Short: "Undo the last action",
		Long: `Reverses the most recent devpod operation.

Each undo creates a new snapshot on the versions branch with
Action: undo, so you never lose history.

Note: landing cannot be undone because changes have already been
pushed to main.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			if list {
				entries := workspace.ListUndoEntries()
				if len(entries) == 0 {
					fmt.Println(format.DimText("Nothing to undo."))
					return nil
				}

				fmt.Println("Undo history (newest first):")
				fmt.Println()

				for i := len(entries) - 1; i >= 0; i-- {
					entry := entries[i]
					timeStr := format.RelativeTime(entry.Timestamp)
					marker := "\u2022"
					if i == len(entries)-1 {
						marker = "\u25b6"
					}
					fmt.Printf("  %s %s  %s\n", marker, entry.Description, format.DimText(timeStr))
				}

				fmt.Println()
				fmt.Println(format.DimText("Run \"devpod undo\" to undo the latest action."))
				return nil
			}

			entry := workspace.GetLastUndoEntry()
			if entry == nil {
				fmt.Println(format.DimText("Nothing to undo."))
				return nil
			}

			feature := workspace.GetCurrentFeature()

			switch entry.Action {
			case "diff-create":
				// Undo diff creation: reverse the commit, take undo snapshot
				gitCmd := exec.Command("git", "reset", "HEAD~1")
				_ = gitCmd.Run()

				// Find and remove the diff from metadata
				if uuid, ok := entry.Data["uuid"].(string); ok {
					if featureSlug, ok := entry.Data["featureSlug"].(string); ok {
						f, err := workspace.LoadFeature(featureSlug)
						if err == nil && f != nil {
							var newDiffs []string
							for _, id := range f.Diffs {
								if id != uuid {
									newDiffs = append(newDiffs, id)
								}
							}
							f.Diffs = newDiffs
							_ = workspace.SaveFeature(*f)
						}
					}
				}

				// Take undo snapshot
				if feature != nil {
					_, _ = takeSnapshot(feature, "", "undo", fmt.Sprintf("Undo: %s", entry.Description))
				}

				_ = workspace.RemoveLastUndoEntry()
				fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)
				fmt.Println(format.DimText("  Your changes are preserved as uncommitted files."))

			case "diff-update":
				// Undo diff update: hard reset to refBefore
				gitCmd := exec.Command("git", "reset", "--hard", entry.RefBefore)
				_ = gitCmd.Run()

				if uuid, ok := entry.Data["uuid"].(string); ok {
					if previousCommit, ok := entry.Data["previousCommit"].(string); ok {
						diff, err := workspace.LoadDiff(uuid)
						if err == nil && diff != nil {
							diff.Commit = previousCommit
							diff.Version = int(math.Max(1, float64(diff.Version-1)))
							_ = workspace.SaveDiff(*diff)
						}
					}
				}

				// Take undo snapshot
				if feature != nil {
					_, _ = takeSnapshot(feature, "", "undo", fmt.Sprintf("Undo: %s", entry.Description))
				}

				_ = workspace.RemoveLastUndoEntry()
				fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)

			case "diff-edit":
				// Undo entering edit mode: hard reset to refBefore, clear editing state
				gitCmd := exec.Command("git", "reset", "--hard", entry.RefBefore)
				_ = gitCmd.Run()

				_ = workspace.SetEditingDiff("")
				_ = workspace.RemoveLastUndoEntry()
				fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)

			case "sync":
				// Undo sync: hard reset to refBefore
				gitCmd := exec.Command("git", "reset", "--hard", entry.RefBefore)
				_ = gitCmd.Run()

				// Take undo snapshot
				if feature != nil {
					_, _ = takeSnapshot(feature, "", "undo", fmt.Sprintf("Undo: %s", entry.Description))
				}

				_ = workspace.RemoveLastUndoEntry()
				fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)
				fmt.Println(format.DimText("  Branch reset to pre-sync state."))

			case "submit":
				// Undo submit: reset diff statuses to draft
				if diffs, ok := entry.Data["diffs"].([]interface{}); ok {
					for _, uuidRaw := range diffs {
						if uuid, ok := uuidRaw.(string); ok {
							diff, err := workspace.LoadDiff(uuid)
							if err == nil && diff != nil {
								diff.Status = "draft"
								_ = workspace.SaveDiff(*diff)
							}
						}
					}
				}

				_ = workspace.RemoveLastUndoEntry()
				fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)
				fmt.Println(format.DimText("  Diff statuses reset to draft."))

			case "land":
				fmt.Println(format.ErrorMsg("Cannot undo a land -- changes have already been pushed to the main codebase."))
				fmt.Println(format.DimText("If needed, revert the change manually on main."))

			default:
				// Generic undo via hard reset
				if entry.RefBefore != "" {
					gitCmd := exec.Command("git", "reset", "--hard", entry.RefBefore)
					_ = gitCmd.Run()

					// Take undo snapshot
					if feature != nil {
						_, _ = takeSnapshot(feature, "", "undo", fmt.Sprintf("Undo: %s", entry.Description))
					}

					_ = workspace.RemoveLastUndoEntry()
					fmt.Printf("%s Undone: %s\n", format.SuccessMsg("\u2713"), entry.Description)
				} else {
					fmt.Println(format.ErrorMsg(fmt.Sprintf("Cannot undo action \"%s\" -- no restore point saved.", entry.Action)))
				}
			}

			return nil
		},
	}

	cmd.Flags().BoolVar(&list, "list", false, "Show what can be undone")

	return cmd
}

// undoTakeSnapshot is called from undo to record the undo action on the versions branch.
func undoTakeSnapshot(feature *workspace.FeatureData, description string) {
	if feature == nil {
		return
	}
	headSHA, _ := git.GetHeadSHA()
	snapshotID := workspace.GetNextSnapshotID(*feature)

	trailers := git.Trailers{
		Snapshot: snapshotID,
		Action:   "undo",
		Feature:  feature.Slug,
		CleanSHA: headSHA,
		Stack:    workspace.GetStackString(*feature),
	}

	vb := feature.VersionsBranch
	if vb == "" {
		vb = workspace.VersionsBranchName(feature.Branch)
	}

	// Only take snapshot if versions branch exists
	if !git.BranchExists(vb) {
		return
	}

	_, _ = git.SnapshotToVersionsBranch(feature.Branch, vb,
		fmt.Sprintf("Undo: %s%s\n", description, git.FormatTrailers(trailers)))

	feature.SnapshotCount++
	_ = workspace.SaveFeature(*feature)

	_ = workspace.SaveUndoEntry(workspace.UndoEntry{
		Action:      "undo-snapshot",
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
		RefBefore:   headSHA,
		Description: description,
		Data:        map[string]interface{}{"snapshotId": snapshotID},
	})
}

File: platform/cli/src/commands/undo.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { execSync } from 'node:child_process';
import * as git from '../git';
import {
  getLastUndoEntry,
  listUndoEntries,
  removeLastUndoEntry,
  loadDiff,
  saveDiff,
  loadFeature,
  saveFeature,
} from '../workspace';
import { errorMessage, relativeTime, diffLabel } from '../format';

export function registerUndoCommand(program: Command): void {
  program
    .command('undo')
    .description('Undo the last action')
    .option('--list', 'Show what can be undone')
    .action(async (opts: { list?: boolean }) => {
      try {
        if (opts.list) {
          const entries = listUndoEntries();
          if (entries.length === 0) {
            console.log(chalk.dim('Nothing to undo.'));
            return;
          }

          console.log(chalk.bold('Undo history (newest first):'));
          console.log();

          for (let i = entries.length - 1; i >= 0; i--) {
            const entry = entries[i]!;
            const time = relativeTime(entry.timestamp);
            const isLatest = i === entries.length - 1;
            const marker = isLatest ? chalk.cyan('\u25b6') : chalk.dim('\u2022');
            console.log(
              `  ${marker} ${entry.description}  ${chalk.dim(time)}`,
            );
          }

          console.log();
          console.log(
            chalk.dim('Run "devpod undo" to undo the latest action.'),
          );
          return;
        }

        const entry = getLastUndoEntry();
        if (!entry) {
          console.log(chalk.dim('Nothing to undo.'));
          return;
        }

        switch (entry.action) {
          case 'diff-create': {
            // Undo diff creation: reset HEAD~1 + remove diff metadata
            execSync('git reset HEAD~1', {
              stdio: 'pipe',
              encoding: 'utf-8',
            });

            // Find and remove the diff from metadata
            // The data should have the diff uuid or we can find the newest one
            const data = entry.data as { uuid?: string; featureSlug?: string };
            if (data.uuid) {
              const diff = loadDiff(data.uuid);
              if (diff && data.featureSlug) {
                const feature = loadFeature(data.featureSlug);
                if (feature) {
                  feature.diffs = feature.diffs.filter(
                    (id) => id !== data.uuid,
                  );
                  saveFeature(feature);
                }
              }
            }

            removeLastUndoEntry();
            console.log(
              `${chalk.green('\u2713')} Undone: ${entry.description}`,
            );
            console.log(
              chalk.dim('  Your changes are preserved as uncommitted files.'),
            );
            break;
          }

          case 'diff-update': {
            // Undo diff update: hard reset to refBefore
            execSync(`git reset --hard ${entry.refBefore}`, {
              stdio: 'pipe',
              encoding: 'utf-8',
            });

            const data = entry.data as {
              uuid?: string;
              previousCommit?: string;
            };
            if (data.uuid && data.previousCommit) {
              const diff = loadDiff(data.uuid);
              if (diff) {
                diff.commit = data.previousCommit;
                diff.version = Math.max(1, diff.version - 1);
                saveDiff(diff);
              }
            }

            removeLastUndoEntry();
            console.log(
              `${chalk.green('\u2713')} Undone: ${entry.description}`,
            );
            break;
          }

          case 'diff-edit': {
            // Undo entering edit mode: hard reset to refBefore, clear editing state
            execSync(`git reset --hard ${entry.refBefore}`, {
              stdio: 'pipe',
              encoding: 'utf-8',
            });

            const { setEditingDiff } = require('../workspace');
            setEditingDiff(null);

            removeLastUndoEntry();
            console.log(
              `${chalk.green('\u2713')} Undone: ${entry.description}`,
            );
            break;
          }

          case 'sync': {
            // Undo sync: hard reset to refBefore
            execSync(`git reset --hard ${entry.refBefore}`, {
              stdio: 'pipe',
              encoding: 'utf-8',
            });

            removeLastUndoEntry();
            console.log(
              `${chalk.green('\u2713')} Undone: ${entry.description}`,
            );
            break;
          }

          case 'submit': {
            // Undo submit: reset diff statuses to draft
            const data = entry.data as {
              diffs?: string[];
            };
            if (data.diffs) {
              for (const uuid of data.diffs) {
                const diff = loadDiff(uuid);
                if (diff) {
                  diff.status = 'draft';
                  diff.updated = new Date().toISOString();
                  saveDiff(diff);
                }
              }
            }

            removeLastUndoEntry();
            console.log(
              `${chalk.green('\u2713')} Undone: ${entry.description}`,
            );
            console.log(
              chalk.dim('  Diff statuses reset to draft.'),
            );
            break;
          }

          case 'land': {
            console.log(
              chalk.red('Cannot undo a land \u2014 changes have already been pushed to the main codebase.'),
            );
            console.log(
              chalk.dim('If needed, revert the change manually.'),
            );
            break;
          }

          default: {
            // Generic undo via hard reset
            if (entry.refBefore) {
              execSync(`git reset --hard ${entry.refBefore}`, {
                stdio: 'pipe',
                encoding: 'utf-8',
              });
              removeLastUndoEntry();
              console.log(
                `${chalk.green('\u2713')} Undone: ${entry.description}`,
              );
            } else {
              console.log(
                chalk.red(`Cannot undo action "${entry.action}" \u2014 no restore point saved.`),
              );
            }
            break;
          }
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
